File: leetcode/101_symmetrictree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def compare(self, p, q):
        if p is None and q is None: return True
        if p is None or q is None: return False
        if p.val != q.val: return False
        
        outer = self.compare(p.left, q.right)
        inner = self.compare(p.right, q.left)
        return outer and inner
    
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None: return True
        return self.compare(root.left, root.right)

File: leetcode/101_symmetrictree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSymmetric(root: TreeNode | null): boolean {
    return sameSame(root, root)
};

function sameSame(p: TreeNode | null, q: TreeNode | null): boolean {
    if (!p && !q) return true
    if (!p || !q) return false
    if (p.val !== q.val) return false

    const sameOutside = sameSame(p.left, q.right)
    const sameInside = sameSame(p.right, q.left)

    return sameOutside && sameInside
}
